src/SimpleJson/SimpleJson.cs

```diff
@@ -1412,6 +1415,21 @@ public virtual object DeserializeObject(object value, Type type)
                             dict.Add(kvp.Key, DeserializeObject(kvp.Value, valueType));
 
                         obj = dict;
+
+#if SIMPLE_JSON_READONLY_COLLECTIONS
+                        // Wrap dictionary in a ReadOnlyDictionary<,>
+                        var genericTypeDefinition = type.GetGenericTypeDefinition();
+                        if (genericTypeDefinition == typeof(IReadOnlyDictionary<,>) ||
+                            genericTypeDefinition == typeof(ReadOnlyDictionary<,>))
+                        {
+                            var ctorType = typeof(IDictionary<,>).MakeGenericType(keyType, valueType);
+                            var genericReadonlyType = typeof(ReadOnlyDictionary<,>).MakeGenericType(keyType, valueType);
+                            var ctor = ReflectionUtils.GetContructor(genericReadonlyType, new Type[] { ctorType });

```

Comment:
Looks like a call to `ConstructorCache[Type]` goes to the `ContructorDelegateFactory` method to get the ctor. That could be a problem, since it looks like it only uses default ctors?